Foundation/NSKeyedArchiver.swift

```diff
@@ -7,183 +7,832 @@
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 
-
+import CoreFoundation
 
 // Archives created using the class method archivedRootDataWithObject used this key for the root object in the hierarchy of encoded objects. The NSKeyedUnarchiver class method unarchiveObjectWithData: will look for this root key as well. You can also use it as the key for the root object in your own archives.
 public let NSKeyedArchiveRootObjectKey: String = "root"
 
+typealias CFKeyedArchiverUID = CFTypeRef
+
+internal let NSKeyedArchiveNullObjectReference = NSKeyedArchiver._createObjectRef(0)
+internal let NSKeyedArchiveNullObjectReferenceName: String = "$null"
+internal let NSKeyedArchivePlistVersion = 100000
+internal let NSKeyedArchiverSystemVersion : UInt32 = 2000
+
+internal func objectRefGetValue(objectRef : CFKeyedArchiverUID) -> UInt32 {
+    assert(objectRef.dynamicType == __NSCFType.self)
+    assert(CFGetTypeID(objectRef) == _CFKeyedArchiverUIDGetTypeID())
+
+    return _CFKeyedArchiverUIDGetValue(unsafeBitCast(objectRef, CFKeyedArchiverUIDRef.self))
+}
+
+internal func escapeArchiverKey(key: String) -> String {
+    if key.hasPrefix("$") {
+        return "$" + key
+    } else {
+        return key
+    }
+}
+
+internal let NSPropertyListClasses : [AnyClass] = [
+    NSArray.self,
+    NSDictionary.self,
+    NSString.self,
+    NSData.self,
+    NSDate.self,
+    NSNumber.self
+]
+
+// NSUniqueObject is a wrapper that allows both hashable and non-hashable objects
+// to be used as keys in a dictionary
+internal struct NSUniqueObject : Hashable {
+    var _backing : Any
+    var _hashValue : () -> Int
+    var _equality : (Any) -> Bool
+    
+    init<T: Hashable>(hashableObject: T) {
+        self._backing = hashableObject
+        self._hashValue = { hashableObject.hashValue }
+        self._equality = {
+            if let other = $0 as? T {
+                return hashableObject == other
+            }
+            return false
+        }
+    }
+    
+    init(_ object: AnyObject) {
+        // FIXME can't we check for Hashable directly?
```

Comment:
May want to document the implications of this (only NSObjects will be uniqued in the object graph because otherwise the pointer itself is used as the hash code).